## Exemplo: raízes da equação do 2º grau

Exemplo:

Calcular as soluções da equação do 2º grau

$a x^2 + b x + c = 0$

ou seja

Dados $a, b$ e $c$, calcular

$x_1 = \frac{-b + \sqrt{b^2 -4 a c}}{2 a}$ e $x_2 = \frac{-b - \sqrt{b^2 -4 a c}}{2 a}$

In [1]:
print 'Este programa calcula x tal que a x2 + b x + c = 0'
# testar com os seguintes valores (1,4,1) , (1,2,1) , (1,1,1)

a = 1
b = 4
c = 1

a = float(a)
b = float(b)
c = float(c)
rdelta = complex(b**2 - 4 * a * c) ** 0.5

x1 = (- b + rdelta) / (2.0*a)
x2 = (- b - rdelta) / (2.0*a)

print "x1 =", x1
print "x2 =", x2

Este programa calcula x tal que a x2 + b x + c = 0
x1 = (-0.267949192431+0j)
x2 = (-3.73205080757+0j)


Correndo o programa com diferentes valores de `a`, `b` e `c`:
```
a = 1
b = 2
c = 1

x1 = (-1+0j)
x2 = (-1+0j)
```

```
a = 1
b = 4
c = 1

x1 = (-0.267949192431+0j)
x2 = (-3.73205080757+0j)
```

```
a = 1
b = 1
c = 1

x1 = (-0.5+0.866025403784j)
x2 = (-0.5-0.866025403784j)
```

O programa funciona e os resultados estão corretos, se considerarmos que os numeros reais podem ser considerados numeros complexos.

No entanto, seria mais adequado que os resultados apresentados pelo programa pudessem **distinguir os casos de soluções complexas dos de soluções reais** apresentando, neste último caso, valores com aspeto de numeros reais:

`-3.73205080757` em vez de `-3.73205080757+0j`.

## Alternativa `if...else`

In [ ]:
# Este programa calcula x tal que a x2 + b x + c = 0
# testar com os seguintes valores (1,4,1) , (1,2,1) , (1,1,1)

a, b, c = 1, 4, 1
a, b, c = float(a), float(b), float(c)

# cálculo do discriminante
discrim = b**2 - 4.0 * a * c

#separar soluções reais das complexas
if discrim < 0.0:
    rdelta = complex(discrim)**0.5
else:
    rdelta = float(discrim)**0.5

x1 = (- b + rdelta) / (2.0 * a)
x2 = (- b - rdelta) / (2.0 * a)

print "x1 =", x1, ", x2 =", x2

**Forma geral das instruções alternativas** `if...else`:

```
if <condição> :
    <comandos para condição verdadeira>
else:
    <comandos para condição falsa>
```

No teste da condição podemos usar:

- `>` maior
- `<` menor
- `>=` maior ou igual
- `<=` menor ou igual
- `==` igual (**são dois sinais de igual consecutivos**)
- `!=` diferente

Correndo o programa com diferentes valores de `a`, `b` e `c`:
```
a = 1
b = 2
c = 1

x1 = -1.0 , x2 = -1.0
```

```
a = 1
b = 4
c = 1

x1 = -0.267949192431 , x2 = -3.73205080757
```

```
a = 1
b = 1
c = 1

x1 = (-0.5+0.866025403784j) , x2 = (-0.5-0.866025403784j)
```

O programa está bem melhor na maneira de apresentar os resultados.

Só falta ter o cuidado de apresentar **um único valor, no caso de uma raíz dupla**.

## Alternativas `if...elif...else`

In [ ]:
# Este programa calcula x tal que a x2 + b x + c = 0
# testar com os seguintes valores (1,4,1) , (1,2,1) , (1,1,1)

a, b, c = 1, 4, 1
a, b, c = float(a), float(b), float(c)

# cálculo do discriminante
discrim = b**2 - 4.0 * a * c

#separar soluções reais das complexas
if discrim < 0.0:
    rdelta = complex(discrim)**0.5
else:
    rdelta = float(discrim)**0.5
x1 = (- b + rdelta) / (2.0 * a)
x2 = (- b - rdelta) / (2.0 * a)

if discrim < 0.0:
    print "2 raízes complexas:"
    print "x1 =", x1
    print "x2 =", x2
elif discrim == 0.0:
    print "1 raíz (dupla):"
    print "x =", x1
else:
    print "2 raízes reais:"
    print "x1 =", x1
    print "x2 =", x2

Podemos ter multiplos comandos na parte do verdadeiro/falso : o **alinhamento** (_identação_) define os blocos

Usando `elif` podemos testar mais do que uma condição.

Correndo o programa com diferentes valores de `a`, `b` e `c`:
```
a = 1
b = 2
c = 1

1 raíz (dupla):
x = -1.0
```

```
a = 1
b = 4
c = 2

2 raízes reais:
x1 = -0.267949192431
x2 = -3.73205080757
```

```
a = 1
b = 1
c = 1

2 raízes complexas:
x1 = (-0.5+0.866025403784j)
x2 = (-0.5-0.866025403784j)
```

## Função `input`

In [1]:
# Este programa calcula x tal que a x2 + b x + c = 0
# testar com os seguintes valores (1,4,1) , (1,2,1) , (1,1,1)

a = float(input('a ='))
b = float(input('b ='))
c = float(input('c ='))

discrim = b**2 - 4.0 * a * c
if discrim < 0.0:
    rdelta = complex(discrim)**0.5
else:
    rdelta = float(discrim)**0.5
x1 = (- b + rdelta) / (2.0 * a)
x2 = (- b - rdelta) / (2.0 * a)

if discrim < 0.0:
    print "2 raízes complexas:"
    print "x1 =", x1
    print "x2 =", x2
elif discrim == 0.0:
    print "1 raíz (dupla):"
    print "x =", x1
else:
    print "2 raízes reais:"
    print "x1 =", x1
    print "x2 =", x2

a =1
b =4
c =1
2 raízes reais:
x1 = -0.267949192431
x2 = -3.73205080757


### Exemplo: Regra dos anos bissextos

#### Se o ano é divisível por 4, então é bissexto

Regra em vigor até 1582 (calendário Juliano, de Júlio César)

In [ ]:
a = input("ano ? ")

if a % 4 == 0 :
    print a , "é bissexto"
else: 
    print a, "nao é bissexto"

Correndo o programa com diferentes valores:
```
ano ? 2015
2015 nao é bissexto
```

```
ano ? 2012
2012 é bissexto
```

```
ano ? 1900
1900 é bissexto
```

```
ano ? 2000
2000 é bissexto
```

No teste da condição podemos usar:

- Conjunção lógica: `and`
- Disjunção lógica: `or`
- Negação: `not`

### Exemplo: Regra dos anos bissextos (calendário moderno)

- Se o ano é divisível por 4, então é bissexto

- Excepto os que são divisíveis por 100: não são bissextos

- Excepto os divisíveis por 100 que sejam exactamente divisíveis por 400: são bissextos.

Regra em vigor após 1582 (calendário Gregoriano, do papa Gregorio XIII)

In [ ]:
a = input("ano ? ")

if a % 4 == 0 and not (a % 100 == 0 and not a % 400 == 0):
    print a , "é bissexto"
else: 
    print a, "nao é bissexto"

Correndo o programa com diferentes valores:
```
ano ? 2015
2015 nao é bissexto
```

```
ano ? 2012
2012 é bissexto
```

```
ano ? 1900
1900 nao é bissexto
```

Reparar que o resultado é diferente para 1900.

```
ano ? 2000
2000 é bissexto
```

Os interessados na história da introdução do calendário gregoriano podem consultar o artigo

[http://en.wikipedia.org/wiki/Gregorian_calendar](http://en.wikipedia.org/wiki/Gregorian_calendar)